File: samples/MicroServices/NBB.Contracts/NBB.Contracts.Api/ServiceCollectionExtensions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using MediatR;
using Microsoft.Extensions.DependencyInjection;

namespace NBB.Contracts.Api
{
    public static class ServiceCollectionExtensions
    {
        public static void AddScopedContravariant<TBase, TResolve>(this IServiceCollection serviceCollection, Assembly assembly = null)
        {
            if (!typeof(TBase).IsGenericType || typeof(TBase).IsOpenGeneric())
                return;

            var baseDescription = typeof(TBase).GetGenericTypeDefinition();
            var baseInnerType = typeof(TBase).GetGenericArguments().First();

            var types = (assembly ?? baseInnerType.Assembly).ScanFor(baseInnerType);
            foreach (var t in types)
                serviceCollection.AddScoped(baseDescription.MakeGenericType(t), typeof(TResolve));
        }

        private static IEnumerable<Type> ScanFor(this Assembly assembly, Type assignableType)
        {
            return assembly.GetTypes().Where(t => assignableType.IsAssignableFrom(t) && !t.IsInterface && !t.IsAbstract);
        }
    }
}

File: samples/MicroServices/NBB.Contracts/NBB.Contracts.Functional.Api/ServiceCollectionExtensions.fs
[<AutoOpen>]
module ServiceCollectionExtensions

open System
open System.Reflection
open MediatR
open Microsoft.Extensions.DependencyInjection

type Assembly with 
    member this.ScanFor (assignableType: Type) =
        this.GetTypes() 
        |> Seq.filter (fun t -> assignableType.IsAssignableFrom(t) && not t.IsInterface &&  not t.IsAbstract)

type IServiceCollection with
    member this.AddScopedContravariant<'TBase, 'TResolve> ?assembly =
        let isInvalid = not typeof<'TBase>.IsGenericType || typeof<'TBase>.IsOpenGeneric()
        if isInvalid then ()
        else  
            let baseDescription = typeof<'TBase>.GetGenericTypeDefinition()
            let baseInnerType = typeof<'TBase>.GetGenericArguments() |> Seq.head
            let searchedAssebly = 
                match assembly with
                | Some assembly -> assembly
                | None -> baseInnerType.Assembly

            let types = searchedAssebly.ScanFor(baseInnerType);
                
            for t in types do
                this.AddScoped(baseDescription.MakeGenericType t, typeof<'TResolve>) |> ignore

            //types |> Seq.iter (fun t -> this.AddScoped(baseDescription.MakeGenericType(t), typeof<'TResolve>) |> ignore)
            //()
            

